### PR TITLE
dev-python/stripe: Version bump to v2.10.1.

### PR DESCRIPTION
**Pull request by seifertm**
* Updated ebuild to EAPI 7
* Added `pypy` and `pypy3` to `PYTHON_COMPAT` (according to [tested versions in tox.ini](https://github.com/stripe/stripe-python/blob/v2.10.1/tox.ini#L7)
* Changed dependency on dev-python/requests (based on [setup.py of v2.10.2](https://github.com/stripe/stripe-python/blob/v2.10.1/setup.py#L46))
* Removed `IUSE="test"` and added `RESTRICT="test"` (see comment)

Closes: https://bugs.gentoo.org/668782

### CODE REVIEW

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+inherit distutils-r1
+
+DESCRIPTION="Stripe python bindings"
+HOMEPAGE="https://github.com/stripe/stripe-python"
+SRC_URI="mirror://pypi/s/${PN}/${P}.tar.gz"
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE=""
```
Comment (hydrapolic) [inline dev-python/stripe/stripe-2.10.1.ebuild]: Since it's empty, you can drop it.

Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: Done!

**Comment by seifertm** [comment]
Removed empty `IUSE` and quoted python implementations with wildcard in `python_gen_cond_dep`.

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+DESCRIPTION="Stripe python bindings"
+HOMEPAGE="https://github.com/stripe/stripe-python"
+SRC_URI="mirror://pypi/s/${PN}/${P}.tar.gz"
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+
+# Tests communicate with an HTTP server
+# see https://github.com/stripe/stripe-python/blob/v2.10.1/README.md#development
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: Why not try to package it, and run it in `python_test`?

Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: I thought tests that communicate over the network are discouraged, because they do not work with `FEATURES="network-sandbox"` which is suggested for [package testing](https://wiki.gentoo.org/index.php?title=Package_testing&oldid=719658).
Or did I get it wrong? I honestly never activated `network-sandbox`.

Comment (mgorny) [reply dev-python/stripe/stripe-2.10.1.ebuild]: Tests that communicate with *outside* network, yes. But you can start the server in test phase, and then it will be local to the emerge process.

Comment (mgorny) [reply dev-python/stripe/stripe-2.10.1.ebuild]: Look at pymongo for a crazy example. Though hopefully yours wouldn't have to be that complex.

Comment (mgorny) [reply dev-python/stripe/stripe-2.10.1.ebuild]: Thanks a lot! Feel free to ping me if you hit any trouble. Given it's golang, @gyakovlev may be of more help though.

Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: Should I file a separate Bug and PR for stripe-mock since it's a new package?

**Comment by mgorny** [comment]
Nah, just add it to this PR.
--
Best regards, 
Michał Górny

**Comment by seifertm** [comment]
Ok, so I assembled an ebuild for `stripe-mock` under `dev-util`. I had virtually no idea what I was doing, because `stripe-mock` is a Golang project and I neither know Go, nor Go ebuilds.
Please review carefully :)

I then implemented `python_test` of `dev-python/stripe` to start the mock server during tests. Similarly to the `pymongo` ebuild, I am trying to start the mock server on different ports until one succeeds, run the tests and tear the server down.

File: dev-util/stripe-mock/metadata.xml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+  <!-- maintainer-needed -->
```
Comment (mgorny) [inline dev-util/stripe-mock/metadata.xml]: Nope, that's not gonna work. The first rule is, never add a package with no maintainer.

File: dev-util/stripe-mock/stripe-mock-0.34.1.ebuild [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# Copyright 1999-2018 Gentoo Foundation
```
Comment (mgorny) [inline dev-util/stripe-mock/stripe-mock-0.34.1.ebuild]: @gyakovlev, could you review this, please?

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+python_test() {
+	declare -i stripe_mock_port=12111
+	local stripe_mock_logfile="${T}/stripe_mock_${EPYTHON}.log"
+	# Try to start stripe-mock until we find a free port
+	while true; do
+		ebegin "Trying to start stripe-mock on port ${stripe_mock_port}"
+		/var/tmp/portage/dev-util/stripe-mock-0.34.1/image/usr/bin/stripe-mock --http-port ${stripe_mock_port} \
+			&> ${stripe_mock_logfile} &
+		local stripe_mock_pid=$(echo $!)
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: Why do you need to `echo` it? oO

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+		sleep 2
+		# Did stripe-mock start?
+		if ps -p ${stripe_mock_pid} &> /dev/null; then
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: This looks like a horribly bad idea, and totally subject to some other process taking over the PID.

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+			einfo "stripe-mock running on port ${stripe_mock_port}"
+			eend 0
+			break
+		else
+			eend 1 "Port ${stripe_mock_port} unavailable"
+			(( stripe_mock_port++ ))
+		fi
+		if [[ ${stripe_mock_port} -gt 65536 ]]; then
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: Do you really plan to try ~50000 times?

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+			die "Unable to start stripe-mock for tests. Please see the logfile located at: ${stripe_mock_logfile}"
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: This message is too long. Aim for <80 chars, use long wrapped `eerror` + short `die` if necessary.

File: dev-python/stripe/stripe-2.10.1.ebuild (hunk 6 above)
Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: You are saying that `$!` could return the PID of a different process due to a race condition.

In this case, I don't know how to kill `stripe-mock` at the end of the tests. The authors of the `pymongo` ebuild use `mongod --dbpath xyz --shutdown`. The mock server here does not provide any such option. Nor does it provide the option to write a pidfile. So I don't known how to get a handle for shutting down the process. Any ideas?

File: dev-util/stripe-mock/metadata.xml (hunk 3 above)
Comment (seifertm) [reply dev-util/stripe-mock/metadata.xml]: @cedk are you will to maintain stripe-mock?

File: dev-python/stripe/stripe-2.10.1.ebuild (hunk 6 above)
Comment (mgorny) [reply dev-python/stripe/stripe-2.10.1.ebuild]: No. What I'm saying is that if the process crashes, the same PID could be taken by another process. We're generally protecting against possibly dangerous behavior here but still, it would be nice to have a better way of determining whether it's running than wait-then-probe. Maybe issue some HTTP request? (wget/curl/whatever)

Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: Didn't even think of this possibility, but it totally makes sense. Issuing a health check via HTTP should be possible. I'll look into it, thanks!

File: dev-util/stripe-mock/metadata.xml (hunk 3 above)
Comment (cedk) [reply dev-util/stripe-mock/metadata.xml]: No, I have no interest in it.

File: dev-python/stripe/stripe-2.10.1.ebuild (hunk 7 above)
Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: I limited it to 10 retries.

File: dev-python/stripe/stripe-2.10.1.ebuild (hunk 6 above)
Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: I am now using curl to send an authenticated request. The ebuild checks whether it receives a successful HTTP response.

File: dev-util/stripe-mock/metadata.xml (hunk 3 above)
Comment (seifertm) [reply dev-util/stripe-mock/metadata.xml]: If this is going to be a blocker, I am willing to be a maintainer for this package.

stripe-mock is planning to add statefulness, at which point I want to use the package for testing myself.

File: dev-util/stripe-mock/stripe-mock-0.34.1.ebuild (hunk 4 above)
Comment (seifertm) [reply dev-util/stripe-mock/stripe-mock-0.34.1.ebuild]: The `golang-build` eclass installs a bunch of `.a` files and all the sources of vendored libraries. Is it supposed to be done this way? I would expect that such things are hidden behind `USE="static-libs"` and `USE="src"`, respectively.

File: dev-util/stripe-mock/metadata.xml [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+<pkgmetadata>
+  <maintainer type="person">
+    <email>m.seifert@digitalernachschub.de</email>
+    <name>Michael Seifert</name>
+  </maintainer>
+  <maintainer type="project">
+    <email>proxy-maint@gentoo.org</email>
+    <name>Proxy Maintainers</name>
+  </maintainer>
+  <longdescription>A mock HTTP server that responds like the real Stripe API.</longdescription>
```
Comment (mgorny) [inline dev-util/stripe-mock/metadata.xml]: That is the same as DESCRIPTION, kill it.

File: dev-util/stripe-mock/stripe-mock-0.34.1.ebuild (hunk 4 above)
Comment (mgorny) [reply dev-util/stripe-mock/stripe-mock-0.34.1.ebuild]: Yes, go is static-only crap. I have no clue about sources but I suppose leave it as-is; @gyakovlev ?

File: dev-util/stripe-mock/stripe-mock-0.34.1.ebuild [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+HOMEPAGE="https://github.com/stripe/stripe-mock"
+SRC_URI="https://github.com/stripe/${PN}/archive/v${PV}.tar.gz -> ${P}.tar.gz"
+KEYWORDS="~amd64 ~x86"
+
+LICENSE="MIT"
+SLOT="0"
+
+src_compile() {
+	GOPATH="${WORKDIR}/${P}:$(get_golibdir_gopath)" \
+		go build -v -work -x ${EGO_BUILD_FLAGS} "${EGO_PN}" -o stripe-mock
```
Comment (mgorny) [inline dev-util/stripe-mock/stripe-mock-0.34.1.ebuild]: Don't you need to `||die`?

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+"
+
+DOCS="LONG_DESCRIPTION.rst CHANGELOG.md README.md"
+
+python_test() {
+	local stripe_mock_port=12111
+	local stripe_mock_retries=10
+	local stripe_mock_logfile="${T}/stripe_mock_${EPYTHON}.log"
+	# Try to start stripe-mock until we find a free port
+	while [[ stripe_mock_retries -gt 0 ]]; do
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: Instead of counting retries you could just compare port.

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+		ebegin "Trying to start stripe-mock on port ${stripe_mock_port}"
+		stripe-mock --http-port ${stripe_mock_port} &> ${stripe_mock_logfile} &
+		local stripe_mock_pid=$!
+		sleep 2
+		# Did stripe-mock start?
+		curl --fail -u "sk_test_123:" \
+			http://localhost:${stripe_mock_port}/v1/customers &> /dev/null
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: I guess it's starting on IPv4, so `127.0.0.1` will be more error-prone.

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+		if [[ $? -eq 0 ]]; then
+			einfo "stripe-mock running on port ${stripe_mock_port}"
+			eend 0
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: I think you want `eend` before other messages. Also, it preserves `$?`, so you can put it above your condition.

File: dev-util/stripe-mock/stripe-mock-0.34.1.ebuild (hunk 10 above)
Comment (seifertm) [reply dev-util/stripe-mock/stripe-mock-0.34.1.ebuild]: It turned out that the build command is wrong. I am now using the default command provided by `golang-build.eclass`

File: dev-util/stripe-mock/stripe-mock-0.34.1.ebuild (hunk 4 above)
Comment (seifertm) [reply dev-util/stripe-mock/stripe-mock-0.34.1.ebuild]: Other ebuilds such as [go-protobuf-1.2.0](https://github.com/gentoo/gentoo/blob/master/dev-go/go-protobuf/go-protobuf-1.2.0.ebuild) seem to remove the bundled sources in `src_install`.

I personally don't like this solution, so I am not going to touch `src_install` for now.

File: dev-python/stripe/stripe-2.10.1.ebuild (hunk 13 above)
Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: The `curl` command is now followed directly by `eend $?`. The messages are logged with either `einfo` in case of success or `eerror` when `$?` is not 0.

File: dev-util/stripe-mock/stripe-mock-0.34.1.ebuild (hunk 10 above)
Comment (seifertm) [reply dev-util/stripe-mock/stripe-mock-0.34.1.ebuild]: Yes, good find!

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+DEPEND="${RDEPEND}
+	test? (
+		>=dev-util/stripe-mock-0.33.0
+		dev-python/pytest[${PYTHON_USEDEP}]
+		dev-python/pytest-mock[${PYTHON_USEDEP}]
+		net-misc/curl
+	)
+"
+
+DOCS="LONG_DESCRIPTION.rst CHANGELOG.md README.md"
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: Array, please.

File: dev-util/stripe-mock/stripe-mock-0.34.1.ebuild (hunk 4 above)
Comment (mgorny) [reply dev-util/stripe-mock/stripe-mock-0.34.1.ebuild]: BTW I find it kinda weird that it installs all that stuff and not the actual executable.

Comment (gyakovlev) [reply dev-util/stripe-mock/stripe-mock-0.34.1.ebuild]: Hi!

it depends not how package is used but generally you don't need it and it's perfectly fine to just `dobin` the binary and accompanying required files (e.g. configs, docs) and be done with it.

`golang-src_install` and the sources/libraries it installs are needed if you your package is a dependency of another go package you are packaging and you want to use system-wide installed go sources instead of temporary bundling stuff during build time and discarding right after.

Same applies to libraries, they are used if you are compiling something against given package, but generally not used at runtime by the package itself.

TLDR: if you don't plan to build any go package that requires `stripe-mock` sources installed system-wide then ditch the sources and probably libraries as well.

Comment (seifertm) [reply dev-util/stripe-mock/stripe-mock-0.34.1.ebuild]: Thanks for the clarification!
I can't foresee the the future, but I find it unlikely that someone will build upon stripe-mock. Therefore, I decided to remove the sources and libraries from the GOPATH directory.

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+		eend $?
+		if [[ $? -eq 0 ]]; then
+			einfo "stripe-mock running on port ${stripe_mock_port}"
+			break
+		else
+			eerror "Port ${stripe_mock_port} unavailable"
+			(( stripe_mock_port++ ))
+		fi
+	done
+	if [[ stripe_mock_port -gt stripe_mock_max_port ]]; then
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: Likewise.

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+"
+
+DOCS=(LONG_DESCRIPTION.rst CHANGELOG.md README.md)
+
+python_test() {
+	local stripe_mock_port=12111
+	local stripe_mock_max_port=12121
+	local stripe_mock_logfile="${T}/stripe_mock_${EPYTHON}.log"
+	# Try to start stripe-mock until we find a free port
+	while [[ stripe_mock_port -le stripe_mock_max_port ]]; do
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: Does this really do what you think it does? Because AFAICS you're comparing two fixed strings.

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+		ebegin "Trying to start stripe-mock on port ${stripe_mock_port}"
+		stripe-mock --http-port ${stripe_mock_port} &> ${stripe_mock_logfile} &
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: Quoting!

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+		sleep 2
+		# Did stripe-mock start?
+		curl --fail -u "sk_test_123:" \
+			http://127.0.0.1:${stripe_mock_port}/v1/customers &> /dev/null
+		eend $?
+		if [[ $? -eq 0 ]]; then
+			einfo "stripe-mock running on port ${stripe_mock_port}"
+			break
+		else
+			eerror "Port ${stripe_mock_port} unavailable"
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: You can pass teh error message to eend, it will be printed only on failure.

File: dev-python/stripe/stripe-2.10.1.ebuild (hunk 16 above)
Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: I couldn't find out why this even works, to be honest :)
Possibly because the `-eq` is for integer comparison only.

All the examples use `[[ "$a" -le "$b" ]]`, so I will stick with syntax that uses variable expansion for the sake of clarity.

File: dev-python/stripe/stripe-2.10.1.ebuild (hunk 17 above)
Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: Good catch! The path to the logfile is now quoted.

File: dev-python/stripe/stripe-2.10.1.ebuild (hunk 18 above)
Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: This gets rid of the "else" branch. Thanks!

File: dev-python/stripe/stripe-2.10.1.ebuild (hunk 16 above)
Comment (mgorny) [reply dev-python/stripe/stripe-2.10.1.ebuild]: Possibly because it ended up as `[[ 0 -le 0 ]]` which is true. I still don't see the fix.

Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: Sorry, messed up the push.

I just uploaded the changes correctly this time.

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+"
+
+DOCS=(LONG_DESCRIPTION.rst CHANGELOG.md README.md)
+
+python_test() {
+	local stripe_mock_port=12111
+	local stripe_mock_max_port=12121
+	local stripe_mock_logfile="${T}/stripe_mock_${EPYTHON}.log"
+	# Try to start stripe-mock until we find a free port
+	while [[ "$stripe_mock_port" -le "$stripe_mock_max_port" ]]; do
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: Always use `${...}` form in ebuilds. Am I really repeating myself all the time?

File: dev-python/stripe/stripe-2.10.1.ebuild [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+	done
+	if [[ "$stripe_mock_port" -gt "$stripe_mock_max_port" ]]; then
+		eerror "Unable to start stripe-mock for tests"
+		die "Please see the logfile located at: ${stripe_mock_logfile}"
+	fi
+
+	STRIPE_MOCK_PORT=${stripe_mock_port} pytest -vv || die "Tests failed with ${EPYTHON}"
+
+	# Tear down stripe-mock
+	kill ${stripe_mock_pid} || die "Unable to stop stripe-mock"
```
Comment (mgorny) [inline dev-python/stripe/stripe-2.10.1.ebuild]: Missing quoting.

Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: I assumed that only things with whitespaces or unknown characters like file paths need to be quoted. There are a few other variable expansions such as `${stripe_mock_port}` should I quote these as well?

Comment (mgorny) [reply dev-python/stripe/stripe-2.10.1.ebuild]: Hmm, I wrongly presumed it might be empty, so nevermind. Feel free to either quote it or leave unquoted.

File: dev-python/stripe/stripe-2.10.1.ebuild (hunk 19 above)
Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: Variable expansions in both ebuilds should now use the `${…}` syntax.

File: dev-python/stripe/stripe-2.10.1.ebuild (hunk 20 above)
Comment (seifertm) [reply dev-python/stripe/stripe-2.10.1.ebuild]: If you notice a potential problem and have to go back while reading the code, others probably will too. I quoted the use of `stripe_mock_pid` here for added clarity.

**Comment by mgorny** [comment]
It looks good. However, one more thing just occurred to me while testing it. Could you move the code starting and stopping `stripe-mock` to `src_test()`, and call `distutils-r1_src_test` in middle of it, with `python_test()` only running the actual tests? This will save us from having to start stripe-mock multiple times for different Python implementations.

**Comment by seifertm** [comment]
Good idea! Especially for users with many Python implementations or if some of the ports are occupied.
I moved the code for starting and stopping `stripe-mock` to `src_test` as you described. I did some tests and it seems to work fine.

I also noticed that `stripe-mock` does not get killed, if you `Ctrl+C` during the tests or something else happens. For example, if the tests fail, the user ends up with a `stripe-mock` process in the background. Is it possible to register a hook that stops the server? Like this:
```
tear_down_stripe_mock() {
    # Stop the server
    …
}

trap tear_down_stripe_mock EXIT
```
